sound/soc/sof/intel/hda.c

```diff
@@ -1201,7 +1334,54 @@ static struct snd_soc_acpi_mach *hda_sdw_machine_select(struct snd_sof_dev *sdev
 			 peripherals->array[i]->id.part_id,
 			 peripherals->array[i]->id.sdw_version);
 
-	return NULL;
+	if (!peripherals->num_peripherals)
+		return NULL;
+
+	/* Create default SDW mach */
+	mach = devm_kzalloc(sdev->dev, sizeof(*mach), GFP_KERNEL);
+	if (!mach)
+		return NULL;
+
+	/* Get link mask and link number */
+	for (i = 0; i < peripherals->num_peripherals; i++)
+		link_mask |= BIT(peripherals->array[i]->bus->link_id);
+
+	link_num = hweight32(link_mask);
+	links = devm_kcalloc(sdev->dev, link_num, sizeof(*links), GFP_KERNEL);
+	if (!links)
+		return NULL;
+
+	link_index = 0;
+	/* Generate snd_soc_acpi_link_adr struct for each peripheral reported by the ACPI table */
+	for (i = 0; i < peripherals->num_peripherals; i++) {
+		links[link_index].adr_d = find_acpi_adr_device(sdev->dev, peripherals->array[i],
+							       &links[link_index], &amp_index);
+
+		/*
+		 * Only increase the link_index when the next periperal is on a different link.
+		 * Assume that peripherals are grouped by link_id in the ACPI table.
+		 */
+		if (i < peripherals->num_peripherals - 1 &&
+		    peripherals->array[i + 1]->bus->link_id != peripherals->array[i]->bus->link_id)
+			link_index++;
+
+		if (link_index > link_num) {
+			dev_err(sdev->dev,
+				"link_index %d exceeds the link number %d link_mask #%x\n",
+				link_index, link_num, link_mask);
+			return NULL;
+		}
+	}
+
+	mach->drv_name = "sof_sdw";
+	mach->sof_tplg_filename = "sof-sdw-generic.tplg";
+	mach->mach_params.links = links;
+	mach->mach_params.link_mask = link_mask;
+	mach->mach_params.platform = dev_name(sdev->dev);
+	mach->get_function_tplg_files = sof_sdw_get_tplg_files;
+	dev_info(sdev->dev, "Use default SDW machine driver %s topology: %s\n",
+		 mach->drv_name, mach->sof_tplg_filename);
```

Comment:
Do we print when we are not using the 'default SDW machine driver sof_sdw toopology: sof-sdw-generic.tplg'?
Do we need to print this in info level?

Reply:
Removed